File: kachery_widgets/widgets/Explore/Explore.js
import React, { Component } from 'react';
import { PythonInterface } from 'reactopya';
import ExploreWidget from './ExploreWidget';
const config = require('./Explore.json');

export default class Explore extends Component {
    static title = 'Explore kachery directories and files'
    static reactopyaConfig = config
    constructor(props) {
        super(props);
        this.state = {
            // javascript state
            path: null,
            
            // python state
            dir_content: null,
            file_content: null,
            status: '',
            status_message: '',

            // other
            pathHistory: []
        }
    }
    componentDidMount() {
        this.pythonInterface = new PythonInterface(this, config);
        this.kacheryManager = new KacheryManager(this);
        this.pythonInterface.start();
        this.setState({
            status: 'started',
            status_message: 'Starting python backend'
        });
        // Use this.pythonInterface.setState(...) to pass data to the python backend
        this.pythonInterface.setState({
            path: this.props.path,
        });
    }
    componentDidUpdate() {

    }
    componentWillUnmount() {
        if (this.pythonInterface) {
            this.pythonInterface.stop();
        }
    }
    _handlePathChanged = (path) => {
        this.setState({
            pathHistory: [...this.state.pathHistory, this.state.path]
        });
        this.pythonInterface.setState({
            path: path
        });
    }
    _handleBackButton = () => {
        let { pathHistory } = this.state;
        if (pathHistory.length === 0) return;
        
        let path0 = pathHistory.pop();

        this.setState({
            pathHistory: pathHistory
        });
        this.pythonInterface.setState({
            path: path0
        });
    }
    render() {
        const { path, dir_content, file_content, pathHistory } = this.state;
        return (
            <ExploreWidget
                path={path || this.props.path}
                dirContent={dir_content}
                fileContent={file_content}
                pathHistory={pathHistory}
                onPathChanged={this._handlePathChanged}
                onBackButton={this._handleBackButton}
                kacheryManager={this.kacheryManager}
            />
        );
    }
}

class KacheryManager {
    constructor(component) {
        this._component = component;
        this._retrievedText = {};
        this._component.pythonInterface.onMessage((msg) => {
            if (msg.name === 'loadedText') {
                this._retrievedText[msg.path] = {
                    text: msg.text,
                    error: msg.error
                };
            }
        });
    }
    async loadText(path) {
        this._component.pythonInterface.sendMessage({
            name: 'loadText',
            path: path
        });
        while (!this._retrievedText[path]) {
            await waitMsec(500);
        }
        return this._retrievedText[path].text;
    }
}

function waitMsec(ms) {
    return new Promise(resolve => setTimeout(resolve, ms));
}

class RespectStatus extends Component {
    state = {}
    render() {
        switch (this.props.status) {
            case 'started':
                return <div>Started: {this.props.status_message}</div>
            case 'running':
                return <div>{this.props.status_message}</div>
            case 'error':
                return <div>Error: {this.props.status_message}</div>
            case 'finished':
                return this.props.children;
            default:
                return <div>Unknown status: {this.props.status}</div>
        }
    }
}

File: kachery_widgets/widgets/Explore/Explore.py
import kachery as ka

class Explore:
    def __init__(self):
        super().__init__()

    def javascript_state_changed(self, prev_state, state):
        ka.set_config(fr='default_readonly', to='')

        self._set_status('running', 'Running Explore')

        self._set_state(
            dir_content=None,
            file_content=None
        )

        print('running', state)
        if state.get('path'):
            try:
                dir_content = ka.read_dir(state['path'])
                file_content = None
            except:
                dir_content = None
                file_content = ka.load_text(state['path'])
            self._set_state(
                dir_content=dir_content,
                file_content=file_content
            )    

        self._set_status('finished', 'Finished Explore ' + state['path'])

    def on_message(self, msg):
        ka.set_config(fr='default_readonly', to='')
        # process custom messages from JavaScript here
        # In .js file, use this.pythonInterface.sendMessage({...})
        if msg['name'] == 'loadText':
            try:
                text = ka.load_text(msg['path'])
            except:
                self._send_message(dict(
                    name='loadedText',
                    path=msg['path'],
                    text=None,
                    error='Problem loading. Perhaps this is not a text file.'
                ))
                return
            self._send_message(dict(
                name='loadedText',
                path=msg['path'],
                text=text
            ))
    
    # Send a custom message to JavaScript side
    # In .js file, use this.pythonInterface.onMessage((msg) => {...})
    def _send_message(self, msg):
        self.send_message(msg)

    # Set the python state
    def _set_state(self, **kwargs):
        self.set_state(kwargs)
    
    # Set error status with a message
    def _set_error(self, error_message):
        self._set_status('error', error_message)
    
    # Set status and a status message. Use running', 'finished', 'error'
    def _set_status(self, status, status_message=''):
        self._set_state(status=status, status_message=status_message)